develop_commit

KonstantinKhmura opened the pull request.

KonstantinKhmura committed:
develop_commit

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/config/AppConfig.java:
```diff
@@ -0,0 +1,6 @@
+package ru.qascooter.praktikumservices.config;
+
+public class AppConfig {
+    public static String APP_URL = "https://qa-scooter.praktikum-services.ru/";
```
Отлично, что урлы вынесены как константы

pomegranate282922 replied:
Не хватает final для константы.

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/pageobject/MainPage.java:
```diff
@@ -0,0 +1,29 @@
+public class MainPage {
+
+    //создаю локатор и метод на аккордеон
+    private static final String accordionItemTag = "accordion__item";
```
Статические константы именуем заглавными буквами через нижнее подчеркивание. Нужно исправить везде

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/pageobject/OrderFirstForm.java:
```diff
@@ -0,0 +1,50 @@
+        WebElement nextButton = webDriver.findElement(nextButtonID);
+        nextButton.click();
+    }
+
```
Методы в page object не делаем статическими. Таким образом ты теряешь все плюшки ооп.
Также нужно сделать поле driver и конструктор в page object, в который передать driver. Это позволит убрать драйвер из всех методов po 
Нужно исправить везде

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/unit/OrderFormLoader.java:
```diff
@@ -0,0 +1,19 @@
+import static ru.qascooter.praktikumservices.config.AppConfig.APP_URL;
+
+public class OrderFormLoader {
+    public static WebDriver Load() {
```
Все методы именуем начиная со строчной буквы. Нужно исправить везде

pomegranate282922 replied:
Не исправлено.

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/unit/OrderFormLoader.java:
```diff
@@ -0,0 +1,19 @@
+        var webDriver = new ChromeDriver();
+        webDriver.get(APP_URL); //открываем тестируемую страницу
+        //webDriver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS); //ставим таймаут на всякий
+        webDriver.findElement(By.className("Button_Button__ra12g")).click();
```
Локаторы должны находиться только в page object, иначе это нарушает page object model.
Нужно исправить

pomegranate282922 replied:
И здесь не исправлено(

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/AccordionMainPageTest.java:
```diff
@@ -0,0 +1,47 @@
+    }
+    @Test
+    public void checkListQuestions() {
+        var mainPage = new MainPage();
```
Такая реализация нарушает page object model. Все локаторы и действия над элементами должны находить в соответствующем page object, а не в тесте.
Нужно исправить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/AccordionMainPageTest.java:
```diff
@@ -0,0 +1,47 @@
+            WebElement expandedBody = webElement.findElement(By.className("accordion__button")); //ищем элемент
+            boolean isExpanded = Boolean.parseBoolean(expandedBody.getAttribute("aria-expanded")) == true;  //получаем свойство атрибута со свойством true
+            assertTrue("Элемент не расскрыт", isExpanded); //проверяем на раскрытие
+        }
```
Нужно проверить, что текст в каждом ответе соответствует ожидаемому.
При этом все проверки должны быть независимыми друг от друга. Нужно сделать или 8 тестовых метола, или использовать параметризацию ( что лучше)
Нужно исправить

pomegranate282922 replied:
Нужно сделать все 8 проверок независимыми. Циклы использовать нельзя - это делает проверки зависимыми.
Нужно исправить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/HeaderOrderButtonTest.java:
```diff
@@ -0,0 +1,35 @@
+    @Test
+    public void whenClickHeaderOrderButton_ThenOrderPageGetsShow() {
+        MainPage.findHeaderOrderButtonAndClick(webDriver); //обращаемся к методу-помошнику из POM
+        var isDisplayed = webDriver.findElement(By.className("Order_Form__17u6u")).isDisplayed(); //получаем свойство isDisplayed
```
Тоже нарушается page object model. Нужно исправить.

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderFirstPageTest.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Parameterized.Parameters // устанавливаю значения для проверок
+    public static Object[][] getUserData() {
+        return new Object[][]{
```
Отлично! Уместно применена параметризация

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderFirstPageTest.java:
```diff
@@ -0,0 +1,76 @@
+        OrderFirstForm.findNextButtonAndClick(webDriver); //обращаемся к методу-помошнику из POM
+
+        //проверка
+        var isDisplayed = webDriver.findElement(By.className("Order_Header__BZXOb")).isDisplayed(); //получаем свойство isDisplayed
```
Здесь нужно убрать локаторы.

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderSecondPageTest.java:
```diff
@@ -0,0 +1,49 @@
+
+        //проверка. на хроме не рабоатет (баг)
+        //проверка
+        var isDisplayed = webDriver.findElement(By.className("Order_ModalHeader__3FDaJ")).isDisplayed(); //получаем свойство isDisplayed
```
Здесь тоже

pomegranate282922 replied:
Нужно проверить, что на странице отображен текст заказ оформлен.
Нужно исправить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderSecondPageTest.java:
```diff
@@ -0,0 +1,49 @@
+        var isDisplayed = webDriver.findElement(By.className("Order_ModalHeader__3FDaJ")).isDisplayed(); //получаем свойство isDisplayed
+        assertTrue("Элемент не отображается", isDisplayed); //проверили
+    }
+}
```
Нет закрытия браузера. Нужно исправить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderSecondPageTest.java:
```diff
@@ -0,0 +1,49 @@
+    public void firstPageIsDone() {
+            OrderFirstPageTest orderFirstPageTest = new OrderFirstPageTest("Иван", "Петров", "Москва", "Черкизовская", "88005553535");
+            var webDriver = OrderFormLoader.Load();
+            orderFirstPageTest.fillFirstForm(webDriver);
```
Вызывать один тест из другого - не лучший вариант. Так они становятся связанными, что неправильно.
В принципе, в задании просят сделать всего два теста на позитивный флоу создания заказа - начинающиеся с разных кнопок. Разбивать на кучу тестов этот один флоу не кажется логичным - по крайней мере, в реализации, когда в итоге все равно один тест вызывает другой. 
нужно исправить.

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/TestSuite.java:
```diff
@@ -0,0 +1,22 @@
+import org.openqa.selenium.chrome.ChromeDriver;
+
+//создаю один тест-сьют для общего запуска всех тестов
+@RunWith(Suite.class)
```
На практике это используется очень редко. FYI.

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderFirstPageTest.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @After
+    public void teardown() {
```
Общие начала и концы во всех тестах лучше вынести в отдельный BaseTest - предок всех тестовых классов. Это уменьшит дублирование кода.
Можно улучшить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderSecondPageTest.java:
```diff
@@ -0,0 +1,49 @@
+    @Test
+    public void firstPageIsDone() {
+            OrderFirstPageTest orderFirstPageTest = new OrderFirstPageTest("Иван", "Петров", "Москва", "Черкизовская", "88005553535");
+            var webDriver = OrderFormLoader.Load();
```
Здесь локальная переменная. Зачем тогда поле в классе?

pomegranate282922 reviewed (commented):
Работа хорошая! Реализована page object model, вынесены урлы как константы, применена параметризация.
Есть и замечания: 
нарушение page object model
вызов одного теста из другого
Тесты на faq зависимы и не проверяют текст ответа 
Методы в po статические 
Нет конструктора в po 
Нет проверки заказ оформлен 
Остальное по мелочи

KonstantinKhmura committed:
Second Commit

Создал два отдельных класса с тестами по заданию: 1 - тест FAQ, 2 - тест заказа с разными наборами данных.

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/config/AppConfig.java:
```diff
@@ -0,0 +1,6 @@
+
+public class AppConfig {
+    public static final String APP_URL = "https://qa-scooter.praktikum-services.ru/";
+    //public static String ORDER_URL = "https://qa-scooter.praktikum-services.ru/order;";
```
Закомментированный код нужно удалить

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/pageobject/OrderSecondForm.java:
```diff
@@ -0,0 +1,53 @@
+    private static final By DROPDOWN_OPTION_ID = By.xpath("//div[@class='Dropdown-option'][text()='сутки']");
+    private static final By COLOR_FIELD_ID = By.xpath("//label[@for='black' and text()='чёрный жемчуг']");
+    private static final By COMMENT_FIELD_ID = By.xpath("//input[@type='text' and @class='Input_Input__1iN_Z Input_Responsible__1jDKN' and @placeholder='Комментарий для курьера']");
+    private static final By ORDER_BUTTON_ID = By.xpath("/html/body/div/div/div[2]/div[3]/button[2]");
```
Такие локаторы - почти абсолютные. Будут ломаться при малейшем изменении верстки. Можно улучшить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/AccordionMainPageTest.java:
```diff
@@ -0,0 +1,69 @@
+        assertTrue("Элемент не расскрыт", isExpanded); //проверяем на раскрытие
+        var displayedText = webElement.findElement(By.className("accordion__panel")).getText();
+        assertTrue("Содержание не соответствует жизненным ожиданиям", displayedText.equals(text));
+
```
Не исправлено - здесь нарушение page object model. Не должно быть локаторов в тесте.
Нужно исправить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderTest.java:
```diff
@@ -0,0 +1,109 @@
+
+        //проверка. на хроме не рабоатет (баг)
+        var isConfirmDisplayed = webDriver.findElement(By.className(".Button_Button__ra12g.Button_Middle__1CSJM")).isDisplayed(); //получаем свойство isDisplayed
+        assertTrue("Элемент не отображается", isConfirmDisplayed); //проверили, что открылось окно с кнопкой "Посомтреть заказ"
```
Этого мало. А если в этом элементе написано - упс, что-то пошло не так? Нужно проверить текст.
Нужно исправить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderTest.java:
```diff
@@ -0,0 +1,109 @@
+        orderSecondForm.setComment("Доставить быстро");  //обращаемся к методу-помошнику из POM
+        orderSecondForm.order(); //обращаемся к методу-помошнику из POM
+        orderSecondForm.confirm(); //обращаемся к методу-помошнику из POM
+    }
```
Везде в тестах есть локаторы. Их быть не должно. Они должны быть описаны в page object, а здесь должны вызываться соответствующие методы из po.
Нужно исправить

pomegranate282922 reviewed (commented):
Многое поправил, супер! Но остались критичные замечания: 
Нарушение page object model
Нет проверки текста успешности создания заказа 
Нейминг немного

KonstantinKhmura committed:
Third Commit

Перевел локаторы в POM, изменил тестирование FAQ. Исправил мелкие ошибки

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/pageobject/MainPage.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public WebElement getAccordionPanel(int index) {
+        var accordion = webDriver.findElement(By.className("accordion"));
```
А зачем эта строчка? Она ничего не делает.

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/pageobject/OrderFirstForm.java:
```diff
@@ -0,0 +1,62 @@
+        this.webDriver = webDriver;
+    }
+
+    public static By getUserStationId(String userStation){
```
Статик лишний

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/AccordionMainPageTest.java:
```diff
@@ -0,0 +1,76 @@
+        webDriver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS); //ставим таймаут на всякий случай
+    }
+    @Test
+    public void checkListQuestions2()
```
Не используем магические цифры в именах. Нужно исправить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/AccordionMainPageTest.java:
```diff
@@ -0,0 +1,76 @@
+            {
+                ((JavascriptExecutor) webDriver).executeScript("arguments[0].scrollIntoView(true);", webElement);
+                webElement.click();
+                WebElement expandedBody = webElement.findElement(By.className("accordion__button"));
```
Не должно быть локаторов в тестах. Они только в page object. Нужно исправить

KonstantinKhmura replied:
Сделал отдельных 8 тестов на FAQ. Время поджимает, а параметризацией с ним у меня не получается. Циклом проверка, которая была - там проверка полей не была взаимосвязана друг с другом. Если находилось несоответствие (я проверял), то цикл шел дальше и в конце выдавал сообщение: в каком поле было не соответствие + ожидаемый и фактический результат. Локаторы, соответственно, из тестов тоже убрал в FAQ

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderTest.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Before
+    public void unit(){
```
Методу нужно дать осмысленное название

pomegranate282922 commented on src/main/java/ru/qascooter/praktikumservices/pageobject/OrderSecondForm.java:
```diff
@@ -0,0 +1,64 @@
+    public void checkOrderIsComplete(){
+        var confirmTextElement = webDriver.findElement(By.className("Order_Text__2broi")); //нашли элемент "Заказ оформлен"
+        var confirmText = confirmTextElement.getText(); //получаем текст
+        assertEquals("Текст не соответствует ожидаемому", "Заказ оформлен", confirmText); //проверили, что заказ Оформлен
```
Эта проверка должна проходить в браузере, отличном от хрома.
Перепроверь, пожалуйста

KonstantinKhmura replied:
Проверил. Проверка действительно не проходила. Изменил локатор к элементу на верный.

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderTest.java:
```diff
@@ -0,0 +1,105 @@
+        var mainPage = new MainPage(webDriver);
+        var orderFirstForm = new OrderFirstForm(webDriver);
+        mainPage.findHeaderOrderButtonAndClick(); //обращаемся к методу-помошнику из POM
+        orderFirstForm.secondFormIsDisplayed();
```
В тесте получилось много ассертов. Это противоречит принципу - один тест одна проверка. 
нужно протестировать флоу - создать заказ ( нажать кнопку заказать, заполнить форму юзера, заполнить форму заказа, проверить, что на странице есть надпись заказ оформлен).
Нужно исправить

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderTest.java:
```diff
@@ -0,0 +1,105 @@
+        //проверка. на хроме не рабоатет (баг)
+        orderSecondForm.checkOrderIsComplete();
+    }
+    public void fillFirstForm(){
```
Вместо того, чтобы делать такие методы в тесте, намного логичнее было бы сделать их шагами в page object. 
можно улучшить

KonstantinKhmura replied:
Перенес все методы в POM, в тестах оставил вызов методов. Была идея сделать методы в POM более обширными, но решил оставить более детализированными.

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/OrderTest.java:
```diff
@@ -0,0 +1,105 @@
+        orderFirstForm.setPhone(userPhone); //обращаемся к методу-помошнику из POM
+        orderFirstForm.clickNextButton(); //обращаемся к методу-помошнику из POM
+    }
+    public void fillSecondForm() {
```
First, second - не самые хорошие имена. Лучше всегда исходить из бизнес логики- userForm, orderForm

pomegranate282922 reviewed (commented):
Опять налицо значительные улучшения в коде.
Но все ещё остаются критичные замечания.
В тесте на faq - нужно сделать независимые (!) проверки: либо 8 разных тестовых метола, либо параметризация. 
также необходимо убрать все локаторы из теста. 
В тесте на создание заказа нужно убрать лишние ассерты и проверить, что тест зелёный в FF, например.
Остальное - по мелочи ( нейминг и пр)

KonstantinKhmura committed:
Fourth Commit

Полностью убрал все локаторы из тестов. Сократил число ассертов до одного в OrderTest. Исправил локатор на проверку оформления заказа (поле с текстом "Заказ оформлен"). Проверил в FF. Убрал методы из тестов - перенес в POM. Объединять не стал ввиду возможной громоздкости.
В тестировании FAQ убрал цикл (хотя при возможном не соответствии этот цикл не падал, завершал тестирование и давал сообщение - где именно несоответствие). Сделал 8 отдельных тестов на каждое поле ввиду ограниченного времени и проблем с оформлением теста через параметризацию: пытался написать код с ней, но ловил много ошибок. Поэтому воспользовался вторым способом из комментария - 8 отдельных тестов.

pomegranate282922 commented on src/test/java/ru/qascooter/praktikum/AccordionMainPageTest.java:
```diff
@@ -0,0 +1,69 @@
+    @Test
+    public void checkEighthFieldOfAccordion (){
+        var mainPage = new MainPage(webDriver);
+        mainPage.checkEigthAccordionField();
```
Да, на этом этапе это возможное решение. Поэтому принимаем! Но сам видишь, насколько много дублирования кода - параметризация решает эти проблемы.
Цикл в том виде как у тебя плох тем, что это самописный велосипед - зачем, если есть отличная возможность параметризованного теста.

KonstantinKhmura replied:
Спасибо. В будущем попробую разобраться.

pomegranate282922 reviewed (approved):
Вот! Теперь то что надо. Это и есть page object model.
коммент оставила про циклы.
